### PR TITLE
More uses of the matches! macro

### PR DESCRIPTION
**Pull request by LingMan**
@rustbot modify labels +C-cleanup +T-compiler

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs
Comment (oli-obk) [inline compiler/rustc_ast/src/token.rs:593]: idk, using guards in `matches!` seems not great for readability to me. I would just keep the current code.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (oli-obk) [inline compiler/rustc_ast_lowering/src/lib.rs:1812]: This can also be 

```suggestion
                    PatKind::Ident(BindingMode::ByValue(Mutability::Mut) | BindingMode::ByRef(Mutability::Mut), ..)
```

rigth?

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_ast/src/ast.rs
Comment (oli-obk) [inline compiler/rustc_ast/src/ast.rs:1947]: I do realize the inconsistency in thinking this code is readable while the other `matches!` is not. I have no good answer for you.

File: compiler/rustc_ast/src/ast.rs
Comment (pickfire) [inline compiler/rustc_ast/src/ast.rs:1097]:
```suggestion
                Some(StmtKind::Expr(_) | StmtKind::Semi(Expr { kind: ExprKind::Ret(_), .. })) => true,
```

I wonder if it is possible to do this?

File: compiler/rustc_ast/src/ast.rs
Comment (pickfire) [reply compiler/rustc_ast/src/ast.rs:1947]: Wow, this is rare. Feels a bit extreme for me.

File: compiler/rustc_ast/src/ast.rs
Comment (oli-obk) [reply compiler/rustc_ast/src/ast.rs:1097]: it is, but it's very noisy. I think we're waiting for pattern aliases before starting to merge nested matches like that.

Comment (pickfire) [reply compiler/rustc_ast/src/ast.rs:1097]: Waiting for pattern aliases?

Comment (oli-obk) [reply compiler/rustc_ast/src/ast.rs:1097]: Well... once we have a way to create pattern aliases in a more convenient way than what `macro_rules` macros can offer, then we can create a pattern alias that allows us to conveniently match on an `Expr` without having to specify all the `kind: Foo, ..` boilerplate.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (LingMan) [reply compiler/rustc_ast_lowering/src/lib.rs:1812]: Yes, that's what I started out with. The line gets too long and the formatter turns it into:
```suggestion
                    PatKind::Ident(
                        BindingMode::ByValue(Mutability::Mut) | BindingMode::ByRef(Mutability::Mut),
                        ..
                    )
```
Wasn't entirely happy with that and ended up with what I proposed. If you'd prefer the four line version, that's also fine with me.

File: compiler/rustc_ast/src/token.rs
Comment (LingMan) [reply compiler/rustc_ast/src/token.rs]: Dropped.

File: compiler/rustc_ast/src/ast.rs
Comment (LingMan) [reply compiler/rustc_ast/src/ast.rs:1097]: Having it split up also gives a good place for the helpful comments on each branch.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (oli-obk) [reply compiler/rustc_ast_lowering/src/lib.rs:1812]: Hmm.. right... Maybe we should just keep the original match then, it is very straightforward.

File: compiler/rustc_ast/src/ast.rs
Comment (LingMan) [reply compiler/rustc_ast/src/ast.rs:1947]: I do get what you're saying, @oli-obk. In fact I was a bit uneasy about that other `matches!` myself but submitted it anyway for consistency and to see what the feedback is. This inconsistency probably means that guards in `matches!`are not by themselves problematic. Certainly something to be careful with though.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (LingMan) [reply compiler/rustc_ast_lowering/src/lib.rs:1812]: There's also the option of:
```suggestion
                    PatKind::Ident(ByValue(Mutability::Mut) | ByRef(Mutability::Mut), ..)
```
which would need `use BindingMode::{ByRef, ByValue};` directly above the hunk.

Comment (oli-obk) [reply compiler/rustc_ast_lowering/src/lib.rs:1812]: Sold :hammer: 

I like it, let's do that

Comment (LingMan) [reply compiler/rustc_ast_lowering/src/lib.rs]: Pushed.

File: compiler/rustc_ast/src/ast.rs
Comment (pickfire) [reply compiler/rustc_ast/src/ast.rs:1097]: How about this?
```suggestion
                Some(StmtKind::Expr(_)) => true,
                Some(StmtKind::Semi(Expr { kind: ExprKind::Ret(_), .. })) => true,
```

Comment (pthariensflame) [reply compiler/rustc_ast/src/ast.rs:1097]: @oli-obk This is the first I've heard about pattern aliases as a wanted feature (other than from me)!  Is there an issue/discussion already?

**Pull request merged**
